pkg/ipmi/ipmi.go

```diff
@@ -269,51 +80,21 @@ func (i *IPMI) RawSendRecv(msg Msg) ([]byte, error) {
 		Data:    unsafe.Pointer(&buf[0]),
 		DataLen: _IPMI_BUF_SIZE,
 	}
-	recv := &recv{
+	recv := &response{
 		addr:    req.addr,
 		addrLen: req.addrLen,
 		msg:     recvMsg,
 	}
 
-	// This will be passed to the RawConn Read function. Until readMsg returns
-	// true, Read blocks and retries until the connection is ready for reading.
-	var result []byte
-	var rerr error
-	readMsg := func(fd uintptr) bool {
-		if err := ioctlGetRecv(fd, _IPMICTL_RECEIVE_MSG_TRUNC, recv); err != nil {
-			rerr = fmt.Errorf("ioctlGetRecv failed with %v", err)
-			return false
-		}
-
-		if recv.msgid != req.msgid {
-			log.Printf("Received wrong message. Trying again.")
-			return false
-		}
-
-		if recv.msg.DataLen >= _IPMI_BUF_SIZE {
-			rerr = fmt.Errorf("data length received too large: %d > %d", recv.msg.DataLen, _IPMI_BUF_SIZE)
-		} else if buf[0] != 0 {
-			rerr = fmt.Errorf("invalid response, expected first byte of response to be 0, got: %v", buf[0])
-		} else {
-			result = buf[:recv.msg.DataLen:recv.msg.DataLen]
-			rerr = nil
-		}
-		return true
-	}
-
-	// Read response.
-	conn, err := i.File.SyscallConn()
+	data, err := i.Handler.ReceiveResponse(req.msgid, recv, buf)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get file rawconn: %v", err)
-	}
-	if err := i.File.SetReadDeadline(time.Now().Add(timeout)); err != nil {
-		return nil, fmt.Errorf("failed to set read deadline: %v", err)
-	}
-	if err := conn.Read(readMsg); err != nil {
-		return nil, fmt.Errorf("failed to read rawconn: %v", err)
+		return nil, err
 	}
+	return data, nil
+
+	// This will be passed to the RawConn Read function. Until readMsg returns
+	// true, Read blocks and retries until the connection is ready for reading.
```

Comment:
Nit: are these comments meant to be kept here? Other than this I don't see other issues, verified on an OCP platform u-root ipmi works.